### ISSUE TITLE
Updates to the upload file endpoint

### ISSUE THREAD

**Issue by rfornea** [opacity/storage-node#102]
Make the following updates to the upload file endpoint:  
-remove `ChunkHash` from request params
-change `FileHash` to `FileHandle`
-don't check for the PartIndex being the 1st part (index 1), instead check that AwsUploadID and AwsObjectKey are nil.  This is because we aren't guaranteed the client will send them in order, so instead we should check whether a multipart upload has been started.  If a multipart upload has not been started, AwsUploadID and AwsObjectKey will be nil.  
-make updates to deal with possible race conditions.  The client is going to bombard us with chunks all at once and will not wait for us to give a success response to the first chunk, so we need to be able to make sure we are not attempting to start a multipart upload that has already been started.  Possible solutions include mutexes or db transactions.  Basically, for every file, there should only be one single chunk for which the check `file.AwsUploadID == nil && file.AwsObjectKey == nil` can be true.

**Comment by pzhao5** [comment]
Fixed the 1) and 2) https://github.com/opacity/storage-node/pull/105

**Comment by pzhao5** [comment]
The PartIndex always start from 1 instead 0??

**Comment by pzhao5** [comment]
I can see the race condition here. You can't get away with it with DB change. I believe Database by default is snapshot isolation which mean write won't block any read for high performance purpose. Adding a lock will lead more deadlock on db in the future and it is hard to debug.

Why not client to send an init request? I think that will better to solve the problem.

**Comment by rfornea** [comment]
> The PartIndex always start from 1 instead 0??

Yeah.  This is an AWS thing.  It will error if we try to use 0.  But we do not have to do the offset, the client will do it before they send it to us.

**Comment by rfornea** [comment]
@funkydrummer  @nullpilot  @pzhao5    

Race condition discussion.  

We can prevent this on the backend, but Philip is concerned it could cause scalability problems in the future.  So we could do the following (1) go ahead and resolve it on the backend, and someday if we ever start having problems (or if we notice problems during performance testing) change it to the frontend waiting for a success response from PartIndex 1 or (2) don't worry about it in the backend and have the frontend wait for a success response from PartIndex 1.  

Thoughts?  I'm open to either.

**Comment by nullpilot** [comment]
The solution with the least friction for now would be using an init request as @pzhao5 suggests. We're making a roundtrip for the metadata already. We could combine the two and start the multipart upload at that point and only return `200` if the metadata was saved _and_ the multipart upload was created, which can happen in parallel.

**Comment by rfornea** [comment]
@nullpilot Sounds good to me.  

@pzhao5 Does this sound okay to you?

**Comment by pzhao5** [comment]
Sounds good to me!  If you look at how aws implement s3 multi part upload.
It also try to avoid this issue too. I don’t think we should try to solve
it cleverly which will lead more pain point down the line.

**Comment by rfornea** [comment]
I'll move this to "done" since we the rest of the issue is finished and we are going with a different approach to address the race condition issue.

**Comment by pzhao5** [comment]
@rfornea  do you open a new issue? What is the API for start upload look like?
